### PR TITLE
[e2e] example of custom environment, with options for the run function

### PR DESCRIPTION
**Pull request by pducolin**
### What does this PR do?

Draft example of custom e2e environments, with an attempt to reduce code required for a non-standard deployment. This is an experiment, requires cleaning before it can be merged.

### Motivation

We are receiving requests for help setting up e2e environments with non standard setup. By standard I mean a host, an agent and a fakeintake, or a host and a agent on docker, plus the fakeintake. There is a perception of too much copy paste in the current implementation, I wonder if this is because we lack documentation for custom environments.

In the second custom run function I use self-exporting components: I will have folks requiring custom environments playing with it and gather feedback.

### CODE REVIEW

File: test/new-e2e/examples/customenv_with_two_vm_test.go [hunk 1]
```diff
@@ -6,49 +6,195 @@
...
+	}
 }
 
-func multiVMEnvProvisioner() e2e.PulumiEnvRunFunc[multiVMEnv] {
+// version two with self-exporting components
+func multiVMEnvProvisionerTwo() e2e.PulumiEnvRunFunc[multiVMEnv] {
 	return func(ctx *pulumi.Context, env *multiVMEnv) error {
+		// first, let's create an aws environment
+		// this is where our remote hosts will live
 		awsEnv, err := aws.NewEnvironment(ctx)
```
Comment (nplanel) [inline test/new-e2e/examples/customenv_with_two_vm_test.go:107]: We probably want to reuse the same `awsEnv` than in `multiVMEnvProvisionerOne()`

File: test/new-e2e/examples/customenv_with_two_vm_test.go [hunk 2]
```diff
@@ -6,49 +6,195 @@
...
+}
+
 type multiVMSuite struct {
 	e2e.BaseSuite[multiVMEnv]
 }
 
 func TestMultiVMSuite(t *testing.T) {
-	e2e.Run(t, &multiVMSuite{}, e2e.WithPulumiProvisioner(multiVMEnvProvisioner(), nil))
+	runFunc := multiVMEnvProvisionerOne
+	if os.Getenv("USE_VARIANT_TWO") != "" {
```
Comment (nplanel) [inline test/new-e2e/examples/customenv_with_two_vm_test.go:194]: not sure this is the best way to run different env, I would prefer to have separated files as today running the same common tests (as today)

File: test/new-e2e/pkg/environments/aws/host/host.go [hunk 3]
```diff
@@ -267,3 +267,43 @@ func Provisioner(opts ...ProvisionerOption) e2e.TypedProvisioner[environments.Ho
...
+	PulumiContext *pulumi.Context
+	// AwsEnvironment is the AWS environment to use.
+	AwsEnvironment *aws.Environment
+	// Options is the set of options for the EC2 VM.
+	Options []ec2.VMOption
+	// Importer is the component that will be used to import the EC2 host from pulumi.
+	Importer *remote.HostOutput
+}
+
+func NewHost(params AwsHostParams) (host *remote.Host, err error) {
```
Comment (nplanel) [inline test/new-e2e/pkg/environments/aws/host/host.go:285]: This kind of function we can reuse everywhere :+1: :)

File: test/new-e2e/pkg/environments/agent/agent.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	CommonEnvironment *config.CommonEnvironment
+	// Host is the host where we want to install the agent.
+	Host *remoteComp.Host
+	// Options is a list of options to configure the agent.
+	Options []agentparams.Option
+	// Importer is the component that will be used to import the agent from pulumi.
+	Importer *agent.HostAgentOutput
+}
+
+func NewAgentOnHost(params HostAgentParams) (hostAgent *agent.HostAgent, err error) {
```
Comment (nplanel) [inline test/new-e2e/pkg/environments/agent/agent.go:30]: we probably need 1 function per environment
`agent.NewHostAgent()`
`agent.NewDockerAgent()`
`agent.NewECSAgent()`
....

File: test/new-e2e/pkg/environments/aws/host/host.go (hunk 3 above)
Comment (clarkb7) [reply test/new-e2e/pkg/environments/aws/host/host.go:285]: 🙌

**Comment by pducolin** [comment]
Replaced by https://github.com/DataDog/datadog-agent/pull/24104

**Pull request closed without merging**